python/tests/test_stats.py

```diff
@@ -653,6 +775,284 @@ def test_empty_ts(self):
+
+    def test_two_tree_time_windows(self):
+        ts = self.get_two_tree_ts()
+        A = self.verify(ts, [[0, 1], [2]], [0], None, [0, ts.max_root_time])
+        assert np.allclose(A, [[[0.6, 0.4]]])
+        A = self.verify(ts, [[0, 1], [2]], [0], None, [0, 1.1, ts.max_root_time])
+        assert np.allclose(A, [[[1.0, 0.0]], [[0.5, 0.5]]])
+        A = self.verify(ts, [[0, 1], [2]], [0], None, [0, 0.5, 1])
+        assert np.allclose(A, [[[np.nan, np.nan]], [[1.0, 0.0]]], equal_nan=True)
+        A = self.verify(ts, [[0, 1], [2]], [0], None, [1, ts.max_root_time, 10])
+        assert np.allclose(A, [[[0.5, 0.5]], [[np.nan, np.nan]]], equal_nan=True)
+
+    def test_one_tree_windows_time_windows(self):
+        ts = tskit.load_text(
+            nodes=io.StringIO(self.small_tree_ex_nodes),
+            edges=io.StringIO(self.small_tree_ex_edges),
+            strict=False,
+        )
+        A = self.verify(ts, [[0, 1], [2, 3, 4]], [0], [0, 1], [0, ts.max_root_time])
+        assert np.allclose(A, [[[[1, 0]]]])
+        A = self.verify(
+            ts, [[0, 1], [2, 3, 4]], [0], [0, 0.2, 1], [0, 1.1, ts.max_root_time]
+        )
+        assert np.allclose(
+            A,
+            [
+                [[[1.0, 0.0]], [[np.nan, np.nan]]],
+                [[[1.0, 0.0]], [[np.nan, np.nan]]],
+            ],
+            equal_nan=True,
+        )
+        A = self.verify(ts, [[0, 1], [2, 3, 4]], [0], [0, 0.2], [0, 0.5, 1])
+        assert np.allclose(A, [[[[1.0, 0.0]], [[np.nan, np.nan]]]], equal_nan=True)
+        A = self.verify(
+            ts, [[0, 1], [2, 3, 4]], [0], [0, 0.2, 1, 1.5], [0, ts.max_root_time, 10]
+        )
+        assert np.allclose(
+            A,
+            [
+                [[[1.0, 0.0]], [[np.nan, np.nan]]],
+                [[[1.0, 0.0]], [[np.nan, np.nan]]],
+                [[[np.nan, np.nan]], [[np.nan, np.nan]]],
+            ],
+            equal_nan=True,
+        )
+
+    def test_two_tree_windows_time_windows(self):
+        ts = self.get_two_tree_ts()
+        A = self.verify(ts, [[0, 1], [2]], [0], [0, 1], [0, ts.max_root_time])
+        assert np.allclose(A, [[[[0.6, 0.4]]]])
+        A = self.verify(ts, [[0, 1], [2]], [0], [0, 0.2, 1], [0, 1.1, ts.max_root_time])
+        assert np.allclose(
+            A,
+            [
+                [[[1.0, 0.0]], [[np.nan, np.nan]]],
+                [[[np.nan, np.nan]], [[0.5, 0.5]]],
+            ],
+            equal_nan=True,
+        )
+        A = self.verify(ts, [[0, 1], [2, 3, 4]], [0], [0, 0.2, 1], [0, 0.5, 1])
+        assert np.allclose(
+            A,
+            [
+                [[[np.nan, np.nan]], [[0.5, 0.5]]],
+                [[[np.nan, np.nan]], [[np.nan, np.nan]]],
+            ],
+            equal_nan=True,
+        )
+
+    def test_span_normalise(self):
+        ts = self.get_two_tree_ts()
+        sample_sets = [[0, 1], [2]]
+        focal = [0]
+        windows = ts.sequence_length * np.random.uniform(size=5)
```

Comment:
Is this OK without a seed?

Reply:
I think it's ok to have some variability?

Reply:
This is actually a general question I have, I thought all tests should be deterministic, but I recall another issue where @jeromekelleher suggested leaving seeds out

Reply:
Well as long as there is no pathological case where things fail, and almost all of the possible combos are a meaningful test then its fine.

Reply:
I vote to set a seed, so that if it *does* fail then we'll be able to debug it. If it's not being adequately tested with a single seed, then we need more tests.

Reply:
All unit tests should have seeds -- the test suite should be fully deterministic.